### PR TITLE
API package updates

### PR DESCRIPTION
**Pull request by pfrazee**
This PR is a collection of updates to the api package that goes along with the lex refactor. I also added a few sugars while I was at it. (By "sugars" I mean shorter methods to wrap the long `app.bsky.feed...` calls.)

- Adds `BskyAgent` which extends `AtpAgent` with a bunch of nice sugar functions.
- Adds a few sugar functions to `AtpAgent`
- Adds `AtUri` as an export of the api package
- Adds `RichText` to the api package which includes...
  - `UnicodeString` for handling the utf8/utf16 indexing
  - Sanitation tooling
  - Facet detection tooling
- Adds a note about string encodings to the lexicon

### CODE REVIEW

File: lexicons/app/bsky/feed/post.json [hunk 1]
```diff
@@ -54,7 +54,7 @@
     },
     "textSlice": {
       "type": "object",
-      "description": "A text segment. Start is inclusive, end is exclusive.",
+      "description": "A text segment. Start is inclusive, end is exclusive. Indices are for utf16-encoded strings.",
```
Comment (dholms) [inline lexicons/app/bsky/feed/post.json:57]: maybe also add a `Deprecated` note on this?

File: packages/api/package.json [hunk 2]
```diff
@@ -14,7 +14,9 @@
   },
   "license": "MIT",
   "dependencies": {
+    "@atproto/uri": "*",
     "@atproto/xrpc": "*",
+    "tlds": "^1.234.0",
```
Comment (dholms) [inline packages/api/package.json:19]: lol that version #

File: packages/api/src/bsky-agent.ts
Comment (dholms) [inline packages/api/src/bsky-agent.ts:1]: cool! this is all done by hand right?

Comment (pfrazee) [reply packages/api/src/bsky-agent.ts:1]: Yep!

File: lexicons/app/bsky/feed/post.json (hunk 1 above)
Comment (pfrazee) [reply lexicons/app/bsky/feed/post.json]: Done

File: packages/api/src/rich-text/detection.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+export async function detectFacets(
+  agent: AtpAgent,
+  text: UnicodeString,
+): Promise<Facet[] | undefined> {
+  let match
+  const facets: Facet[] = []
+  {
+    // mentions
+    const re = /(^|\s|\()(@)([a-zA-Z0-9.-]+)(\b)/g
```
Comment (dholms) [inline packages/api/src/rich-text/detection.ts:16]: we actually have a handle regex now over in the identifiers package if you wanna use that 👀

Comment (dholms) [reply packages/api/src/rich-text/detection.ts:16]: oh it doesn't take into account the leading `@` tho

**Review by dholms — APPROVED** [review]
Nice this looks great to me 🙌 

The in-comment docs are quite nice

File: packages/api/src/bsky-agent.ts [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    this.api.app.bsky.feed.getAuthorFeed(params, opts)
+
+  getPostThread: typeof this.api.app.bsky.feed.getPostThread = (params, opts) =>
+    this.api.app.bsky.feed.getPostThread(params, opts)
+
+  getPost: typeof this.api.app.bsky.feed.post.get = (params) =>
+    this.api.app.bsky.feed.post.get(params)
+
+  getLikes: typeof this.api.app.bsky.feed.getLikes = (params, opts) =>
+    this.api.app.bsky.feed.getLikes(params, opts)
```
Comment (devinivy) [inline packages/api/src/bsky-agent.ts:27]: Definitely not too hung-up on it, but due to the use of the sugared `post.get()` I think that `getPost` may have unwrapped the result out of `response.data`, whereas `getLikes` and others return the whole `XRPCResponse` object.

File: packages/api/src/rich-text/detection.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    while ((match = re.exec(text.utf16))) {
+      if (!isValidDomain(match[3]) && !match[3].endsWith('.test')) {
+        continue // probably not a handle
+      }
+
+      const start = text.utf16.indexOf(match[3], match.index) - 1
+      facets.push({
+        $type: 'app.bsky.richtext.facet',
```
Comment (devinivy) [inline packages/api/src/rich-text/detection.ts:20]: I can't think of anywhere the `$type` is strictly necessary on facets yet (is there one perhaps I've forgotten?), but either way I see no issue including it.

File: packages/api/src/rich-text/rich-text.ts [hunk 6]
```diff
@@ -0,0 +1,359 @@
+/*
+= Rich Text Manipulation
```
Comment (devinivy) [inline packages/api/src/rich-text/rich-text.ts:2]: This is beautiful.

File: packages/api/src/rich-text/rich-text.ts [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+    }
+  }
+  return facets
+}
+
+function cloneDeep<T>(v: T): T {
+  if (typeof v === 'undefined') {
+    return v
+  }
+  return JSON.parse(JSON.stringify(v))
```
Comment (devinivy) [inline packages/api/src/rich-text/rich-text.ts:358]: Not sure if it fits the RN environment but there's also `structuredClone()` for this, which is pretty friendly.

File: packages/api/src/rich-text/unicode.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    return this.utf8.byteLength
+  }
+
+  slice(start?: number, end?: number): string {
+    return decoder.decode(this.utf8.slice(start, end))
+  }
+
+  utf16IndexToUtf8Index(i: number) {
+    return encoder.encode(this.utf16.slice(0, i)).byteLength
+  }
```
Comment (devinivy) [inline packages/api/src/rich-text/unicode.ts:32]: 💎

**Review by devinivy — APPROVED** [review]
Excellent stuff!  Excited to use this for tests on the pds, etc.

**Pull request merged**